ipam/allocator.go

```diff
@@ -159,97 +187,293 @@ const (
-	num := len(internalSubnetList)
-	ones, bits := internalSubnetList[0].Mask.Size()
-	numAddresses := 1 << uint(bits-ones)
+// ReleasePool releases the address pool identified by the passed id
+func (a *Allocator) ReleasePool(poolID string) error {
+	var (
+		p, pp *PoolData
+		rk    *subnetKey
+		ok    bool
+	)
+	k := subnetKey{}
+	if err := k.FromString(poolID); err != nil {
+		return types.BadRequestErrorf("invalid pool id: %s", poolID)
+	}
 
-	for i := 0; i < num; i++ {
-		smallKey := subnetKey{parentKey.addressSpace, parentKey.subnet, internalSubnetList[i].String()}
-		limit := uint32(numAddresses)
+retry:
+	a.Lock()
+	p, ok = a.subnets[k]
+	if !ok {
+		a.Unlock()
+		return ipamapi.ErrBadPool
+	}
+	p.RefCount--
+	if p.RefCount == 0 {
+		delete(a.subnets, k)
+		rk = &k
+	}
 
-		if ipVer == v4 && i == num-1 {
-			// Do not let broadcast address be reserved
-			limit--
+	if p.Range != nil {
+		pp, ok = a.subnets[p.ParentKey]
+		if !ok {
+			a.Unlock()
+			return types.InternalErrorf("cannot find parent pool for %s", poolID)
 		}
-
-		// Generate the new address masks. AddressMask content may come from datastore
-		h, err := bitseq.NewHandle(dsDataKey, a.getStore(), smallKey.String(), limit)
-		if err != nil {
-			return err
+		pp.RefCount--
+		if pp.RefCount == 0 {
+			delete(a.subnets, p.ParentKey)
+			rk = &p.ParentKey
 		}
-
-		if ipVer == v4 && i == 0 {
-			// Do not let network identifier address be reserved
-			h.Set(0)
+		a.Unlock()
+		return nil
+	}
+	a.Unlock()
+	if err := a.writeToStore(); err != nil {
+		if _, ok := err.(types.RetryError); !ok {
+			return types.InternalErrorf("pool (%s) removal failed because of %v", poolID, err)
 		}
+		// Update to latest
+		if erru := a.readFromStore(); erru != nil {
+			return fmt.Errorf("failed to get updated pool config from datastore (%v) after (%v)", erru, err)
+		}
+		goto retry
+	}
 
-		a.Lock()
-		a.addresses[smallKey] = h
-		a.Unlock()
+	if rk == nil {
+		return nil
 	}
+
+	a.Lock()
+	if bm, ok := a.addresses[*rk]; ok {
+		bm.Destroy()
+	}
+	delete(a.addresses, *rk)
+	a.Unlock()
+
 	return nil
 }
 
+func (a *Allocator) getPredefineds(as string) []*net.IPNet {
+	a.Lock()
+	defer a.Unlock()
+	l := make([]*net.IPNet, 0, len(a.predefined[as]))
+	for _, pool := range a.predefined[as] {
+		l = append(l, pool)
+	}
+	return l
+}
+
+func (a *Allocator) getPredefinedPool(as string, ipV6 bool) (string, *net.IPNet, error) {
+	var v ipVersion
+	v = v4
+	if ipV6 {
+		v = v6
+	}
+
+	if as != localAddressSpace && as != globalAddressSpace {
+		return "", nil, fmt.Errorf("no default pool availbale for non-default addresss spaces")
+	}
+
+	for _, nw := range a.getPredefineds(as) {
+		if v != getAddressVersion(nw.IP) {
+			continue
+		}
+		cw := types.GetIPNetCanonical(nw)
+		if _, ok := a.subnets[subnetKey{addressSpace: as, subnet: cw.String()}]; ok {
+			continue
+		}
+		if !a.contains(as, cw) {
+			if err := netutils.CheckRouteOverlaps(nw); err == nil {
```

Comment:
Yes Thanks.
We agreed that the os overlap check must be done only over the local address space.